lib/Parse/ParseExpr.cpp

```diff
@@ -2905,9 +2910,67 @@ ExprResult Parser::ParseBoundsExpression() {
     SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
 
   PT.consumeClose();
+
+  if (StartsRelativeBoundsClause(Tok) && !Result.isInvalid()) {
```

Comment:
It is  problematic to just call `SkipUntil` if Result is invalid because of a typechecking error.  `SkipUntil` does not handle nested parentheses correctly and could produce an incorrect syntax error.   I think this is a general issue with `SkipUntil` in clang.

Consider code of the form:
`bounds(e1, e2) rel_align_value((sizeof(T))`, where the programmer put an extra parenthesis around the body of the rel_align_value constant and `e1` or `e2` has a typechecking error.